### ISSUE TITLE
Separate RHS and state vectors for `FunctionOperator`

### ISSUE THREAD

**Issue by danielwe** [SciML/SciMLOperators.jl#57]
Currently, `FunctionOperator` can wrap functions on the form `f(u, p, t)` representing the application of an operator `L(u, p, t)` to the vector `u`, i.e., the operation `L(u, p, t) * u`. However, it should be possible to apply `L(u, p, t)` to other vectors than `u`, i.e., operations `L(u, p, t) * v`. As far as I can tell, this is possible for all operator types except `FunctionOperator`. Perhaps it would make more sense for the function wrapped by `FunctionOperator` to have signature `f(v, u, p, t)` (or `f!(out, v, u, p, t)` for in-place)?

Or, in combination with #48, the supported signatures would be `f(v, u, p, t, α, β)` and `f!(out, v, u, p, t, α, β)` (modulo your preferred argument order).

**Comment by vpuri3** [comment]
what you could do is pass the state vector `u` via the `p` parameter, and apply the operator to some other vector `v`,
```julia
update_coefficients!(L, u, p, t) = update_coefficients!(L, p[1], p[2], t)
w = L(v, (u, p_old,), t) # w = L(u, p_old, t) * v
```

**Comment by danielwe** [comment]
Fair enough. Actually redefining `update_coefficients!` is probably not meaningful, but replacing `p` with `(u, p)` and updating your function to work with that would do the job.

However, it's a bit of a jarring break with the very consistent `(u, p, t)` interface across the rest of the ecosystem, where `u` and `p` always refer to the same thing. It feels like if the signature for an in-place jacobian were `jac!(J, p, t)` and you had to use `p` as a vehicle if you wanted `J` to depend on `u`.

If this is going to be the sanctioned way to separate the state that your operator depends on from the vector it acts on, there should probably be an example in the docs to illustrate how to do it.

**Comment by danielwe** [comment]
No actually, it doesn't work, because `update_coefficients!` is called recursively on the nodes in the operator tree, so the arguments have to have the same meaning everywhere. The only way for the user to customize how `FunctionOperator` is updated would be to redefine its `update_coefficients!` method, which would be type piracy and wouldn't be able to differentiate between different instances (the alternative would be for `FunctionOperator` to accept a `update_func` argument, but that's messy because it would require the user to deal with the internals of the struct).

Hence I maintain that the appropriate signatures for the function wrapped by `FunctionOperator` are `f(v, u, p, t)` and `f!(w, v, u, p, t)`. There needs to be a distinction between the argument the function acts on as an operator and the ones it depends on as parameters, and `u` is always one of the latter (even if it _can_ also be the former). You should be able to do the following for `FunctionOperator` and get the result you expect, just like you can for any other operator:

```
update_coefficients!(L, u, p, t)
w = L * v
```

(Of course, calling the operator as a function should dispatch to a call where `u` plays both roles: `L(u, p, t) = f(u, u, p, t)`, equivalent to how it works for every other operator).

**Comment by vpuri3** [comment]
does something like this work?
```julia
    function Ajac(Jv, v, u, p, t;A=A)
        SciMLOperators.update_coefficients!(A, u, p, t)
        mul!(Jv, A, v)
    end
```
where `A isa SciMLOperator`

**Comment by ChrisRackauckas** [comment]
Fixed in the v1.0
